libs/crypto/proof_of_work.go

```diff
@@ -0,0 +1,111 @@
+package crypto
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"log"
+	"math"
+	"math/big"
+)
+
+const (
+	Sha3     = "sha3_24_rounds"
+	maxNonce = math.MaxInt64
+)
+
+var prefix = []byte("Vega_SPAM_PoW")
+
+// PoW calculates proof of work given block hash, transaction hash, target difficulty and a hash function.
+// returns the nonce, the hash and th error if any.
+func PoW(blockHash string, txID string, difficulty int, hashFunction string) (int, []byte, error) {
+	var hashInt big.Int
+	var h []byte
+	var err error
+	nonce := 0
+	target := big.NewInt(1)
+	target.Lsh(target, uint(256-difficulty))
+
+	if len(txID) < 1 {
+		return nonce, h, errors.New("transaction ID cannot be empty")
+	}
+
+	if len(blockHash) != 64 {
+		return nonce, h, errors.New("incorrect block hash")
+	}
+
+	for nonce < maxNonce {
+		data := prepareData(blockHash, txID, nonce)
+		h, err = hash(data, hashFunction)
+		if err != nil {
+			return nonce, h, err
+		}
+
+		if nonce%1000 == 0 {
+			println(nonce)
+		}
+
+		hashInt.SetBytes(h[:])
+		if hashInt.Cmp(target) == -1 {
+			break
+		} else {
+			nonce++
+		}
+	}
+
+	return nonce, h[:], nil
+}
+
+// Verify checks that the hash with the given nonce results in the target difficulty.
+func Verify(blockHash string, txID string, nonce int, hashFuncion string, difficulty int) bool {
```

Comment:
Should `difficulty` be a `uint`? Feel like we need some checks on `difficulty` too, it panics if you pass in `257`.